contrib/lkt/language/parser.py

```diff
@@ -2208,6 +2208,41 @@ def expected_type_predicate(expected_type=T.TypeDecl.entity):
     def invalid_expected_type_error_name():
         return S("an array")
 
+    @langkit_property()
+    def expr_context_free_type():
+        """
+        Infer type from array elements context_free_type.
+        """
+        # Get the context_free_type of the array elements if any
+        types = Var(
+            Entity.exprs.filter(
+                # Filter out Lit expressions to avoid infinite recursion
+                lambda e: Not(
+                    Or(e.is_a(T.NumLit),
+                       e.is_a(T.CharLit),
+                       e.is_a(T.StringLit))
+                )
+            ).filtermap(
+                # Keep only elements context_free_types
```

Comment:
Keep non-null context-free types only*

But thinking more about this, I think this is an architecture smell: first, we need to keep here the list of nodes for which `expr_context_free` may end up calling back `ArrayLiteral`’s, it’s a code maintainance issue.

But maybe more importantly, it feels wrong that `Lit.expr_context_free_type` does checks the context (i.e. info coming from its parents) in order to compute the type. What about moving the code currently in `Lit.expr_context_free_type_impl` (which checks the context) to `Expr.array_element_type`? (as this one already checks the context, through its call to `.expected_type`) This way we can get rid of the above type-based filtering and preserve the meaning of “context-free type”.

Reply:
I agree, I was not feeling 100% sure of me on that point. Will rework this part.